### PR TITLE
Add tests for microtask checkpoints and custom element constructors

### PR DESCRIPTION
**Pull request by domenic**
Follows https://github.com/whatwg/html/pull/2457 and https://github.com/whatwg/dom/pull/430.

/cc @rniwa

---
This change is [Reviewable](https://reviewable.io/reviews/w3c/web-platform-tests/5208)

### CODE REVIEW

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/214125978)

# Firefox (nightly channel) #
Testing web-platform-tests at revision b051dc4396031e850c1f74c0ce6ae2c99d385c67
Using browser at version BuildID 20170321110237; SourceStamp ca4ae502156eaea6fffb296bb9c3b3930af8ab58
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5208/custom-elements/microtasks-and-constructors.html">/custom-elements/microtasks-and-constructors.html</a></summary>


|                                                         Subtest                                                          | Results |                                               Messages                                               |
|--------------------------------------------------------------------------------------------------------------------------|---------|------------------------------------------------------------------------------------------------------|
|                                                                                                                          | OK      |                                                                                                      |
| `Microtasks evaluate immediately when the stack is empty inside the parser`                                              | FAIL    | `assert_array_equals: lengths differ, expected 2 got 0`                                              |
| `Microtasks evaluate immediately when the stack is empty inside the parser, causing the checks on no attributes to fail` | FAIL    | `assert_true: The element must be a HTMLUnknownElement expected true got false`                      |
| `Microtasks evaluate afterward when the stack is not empty using createElement()`                                        | FAIL    | `assert_array_equals: lengths differ, expected 3 got 1`                                              |
| `Microtasks evaluate afterward when the stack is not empty using the constructor`                                        | FAIL    | `assert_array_equals: property 1, expected "promise microtask" but got "MutationObserver microtask"` |
| `Microtasks evaluate afterward when the stack is not empty due to upgrades`                                              | FAIL    | `assert_array_equals: lengths differ, expected 3 got 1`                                              |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/214125979)

# Chrome (unstable channel) #
Testing web-platform-tests at revision b051dc4396031e850c1f74c0ce6ae2c99d385c67
Using browser at version 59.0.3047.0 dev
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5208/custom-elements/microtasks-and-constructors.html">/custom-elements/microtasks-and-constructors.html</a></summary>


|                                                         Subtest                                                          | Results | Messages |
|--------------------------------------------------------------------------------------------------------------------------|---------|----------|
|                                                                                                                          | OK      |          |
| `Microtasks evaluate immediately when the stack is empty inside the parser`                                              | PASS    |          |
| `Microtasks evaluate immediately when the stack is empty inside the parser, causing the checks on no attributes to fail` | PASS    |          |
| `Microtasks evaluate afterward when the stack is not empty using createElement()`                                        | PASS    |          |
| `Microtasks evaluate afterward when the stack is not empty using the constructor`                                        | PASS    |          |
| `Microtasks evaluate afterward when the stack is not empty due to upgrades`                                              | PASS    |          |

</details>
</details>

**Comment by w3c-bots** [comment]
These tests are now available on [w3c-test.org](https://w3c-test.org/submissions/5208/)

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/214127216)

# Firefox (nightly channel) #
Testing web-platform-tests at revision 8db354b68ae821f1ed09fa8a1ebd72ab2e14f0f1
Using browser at version BuildID 20170321110237; SourceStamp ca4ae502156eaea6fffb296bb9c3b3930af8ab58
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5208/custom-elements/microtasks-and-constructors.html">/custom-elements/microtasks-and-constructors.html</a></summary>


|                                                         Subtest                                                          | Results |                                               Messages                                               |
|--------------------------------------------------------------------------------------------------------------------------|---------|------------------------------------------------------------------------------------------------------|
|                                                                                                                          | OK      |                                                                                                      |
| `Microtasks evaluate immediately when the stack is empty inside the parser`                                              | FAIL    | `assert_array_equals: lengths differ, expected 2 got 0`                                              |
| `Microtasks evaluate immediately when the stack is empty inside the parser, causing the checks on no attributes to fail` | FAIL    | `assert_true: The element must be a HTMLUnknownElement expected true got false`                      |
| `Microtasks evaluate afterward when the stack is not empty using createElement()`                                        | FAIL    | `assert_array_equals: lengths differ, expected 3 got 1`                                              |
| `Microtasks evaluate afterward when the stack is not empty using the constructor`                                        | FAIL    | `assert_array_equals: property 1, expected "promise microtask" but got "MutationObserver microtask"` |
| `Microtasks evaluate afterward when the stack is not empty due to upgrades`                                              | FAIL    | `assert_array_equals: lengths differ, expected 3 got 1`                                              |

</details>
</details>

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/214127217)

# Chrome (unstable channel) #
Testing web-platform-tests at revision 8db354b68ae821f1ed09fa8a1ebd72ab2e14f0f1
Using browser at version 59.0.3047.0 dev
Starting 10 test iterations
All results were stable
## All results ##
<details>
<summary>1 test ran</summary>
<details>
<summary><a href="http://w3c-test.org/submissions/5208/custom-elements/microtasks-and-constructors.html">/custom-elements/microtasks-and-constructors.html</a></summary>


|                                                         Subtest                                                          | Results | Messages |
|--------------------------------------------------------------------------------------------------------------------------|---------|----------|
|                                                                                                                          | OK      |          |
| `Microtasks evaluate immediately when the stack is empty inside the parser`                                              | PASS    |          |
| `Microtasks evaluate immediately when the stack is empty inside the parser, causing the checks on no attributes to fail` | PASS    |          |
| `Microtasks evaluate afterward when the stack is not empty using createElement()`                                        | PASS    |          |
| `Microtasks evaluate afterward when the stack is not empty using the constructor`                                        | PASS    |          |
| `Microtasks evaluate afterward when the stack is not empty due to upgrades`                                              | PASS    |          |

</details>
</details>

File: custom-elements/microtasks-and-constructors.html [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+<x-bad></x-bad>
+
+<script>
+"use strict";
+
+test(() => {
+  const xBad = document.querySelector("x-bad");
+  assert_false(xBad.hasAttribute("attribute"), "The attribute must not be present");
+  assert_true(xBad instanceof HTMLUnknownElement, "The element must be a HTMLUnknownElement");
+}, "Microtasks evaluate immediately when the stack is empty inside the parser, causing the " +
```
Comment (foolip) [inline custom-elements/microtasks-and-constructors.html]: I don't understand this description, is it in reference to a check somewhere in https://html.spec.whatwg.org/multipage/scripting.html#concept-upgrade-an-element?

AFAICT, the test is either asserting that the `doMicrotasks` bit above hasn't run yet?

Comment (domenic) [inline custom-elements/microtasks-and-constructors.html]: It's referring to the last step in https://html.spec.whatwg.org/#clean-up-after-running-script.

**Comment by w3c-bots** [comment]
[View the complete job log.](https://travis-ci.org/w3c/web-platform-tests/jobs/225393567)

# Lint #

Passed

File: custom-elements/microtasks-and-constructors.html (hunk 1 above)
Comment (foolip) [inline custom-elements/microtasks-and-constructors.html]: OK, so the microtask is in fact run, the attribute is added, then there's some "checks on no attributes" that I guess happens as part of the upgrade. I couldn't find an explicit attribute-related check anywhere.

I played a bit with the test and figured out that I don't understand custom elements. The `this` in in `this.setAttribute` isn't the same thing as `xBad` is later. I suppose that's only natural, one can't turn an instance of HTMLUnknownElement into an instance of something else.

Some links to help guide the less familiar reader (me) might be nice, but I'll merge this and that can be an optional follow-up.

**Pull request merged**